src/oskbd/linux.rs

```diff
@@ -70,79 +65,33 @@ impl KbdIn {
 }
 
 pub struct KbdOut {
-    device: File,
+    device: UInputDevice,
 }
 
 impl KbdOut {
     pub fn new() -> Result<Self, io::Error> {
-        let mut uinput_out_file = OpenOptions::new().write(true).open("/dev/uinput")?;
-
-        unsafe {
-            let rc = uinput_sys::ui_set_evbit(uinput_out_file.as_raw_fd(), uinput_sys::EV_SYN);
-            if rc != 0 {
-                log::error!("ui_set_evbit for EV_SYN returned {}", rc);
-                return Err(io::Error::new(
-                    io::ErrorKind::Other,
-                    "ui_set_evbit failed for EV_SYN",
-                ));
-            }
-            let rc = uinput_sys::ui_set_evbit(uinput_out_file.as_raw_fd(), uinput_sys::EV_KEY);
-            if rc != 0 {
-                log::error!("ui_set_evbit for EV_KEY returned {}", rc);
-                return Err(io::Error::new(
-                    io::ErrorKind::Other,
-                    "ui_set_evbit failed for EV_KEY",
-                ));
-            }
-
-            for key in 0..300 {
-                let rc = uinput_sys::ui_set_keybit(uinput_out_file.as_raw_fd(), key);
-                if rc != 0 {
-                    log::error!("ui_set_keybit for {} returned {}", key, rc);
-                    return Err(io::Error::new(io::ErrorKind::Other, "ui_set_keybit failed"));
-                }
-            }
-
-            let mut uidev: uinput_user_dev = mem::zeroed();
-
-            const PROG_NAME: &[u8] = "kanata".as_bytes();
-            let copy_len = std::cmp::min(PROG_NAME.len(), uidev.name.len());
-            assert!(copy_len <= uidev.name.len());
-            for (i, c) in PROG_NAME.iter().copied().enumerate().take(copy_len) {
-                uidev.name[i] = c as c_char;
-            }
-
-            uidev.id.bustype = 0x3; // BUS_USB
-            uidev.id.vendor = 0x1;
-            uidev.id.product = 0x1;
-            uidev.id.version = 1;
-
-            let uidev_bytes =
-                slice::from_raw_parts(mem::transmute(&uidev), mem::size_of::<uinput_user_dev>());
-            uinput_out_file.write_all(uidev_bytes)?;
-            let rc = uinput_sys::ui_dev_create(uinput_out_file.as_raw_fd());
-            if rc != 0 {
-                log::error!("ui_dev_create for returned {}", rc);
-                return Err(io::Error::new(io::ErrorKind::Other, "ui_dev_create failed"));
-            }
+        let device = UninitDevice::new()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "UninitDevice::new() failed"))?;
+
+        device.set_name("kanata");
+        device.set_bustype(BusType::BUS_USB as u16);
+        device.set_vendor_id(0x1);
+        device.set_product_id(0x1);
+        device.set_version(1);
+
+        device.enable(&EventType::EV_SYN)?;
+        device.enable(&EventType::EV_KEY)?;
+        // TODO find a better way to iter enum EV_KEY, i.e. not hardcode the upper bound
+        for key in (0..750).into_iter().filter_map(enums::int_to_ev_key) {
```

Comment:
Do you have any ideas about this?

Reply:
I believe evdev defines a constant stating what the max value is, **but** I think we should actually limit this to `300` for now (as it was before). There was a weird Wayland bug where kanata wouldn't work when using too many values in the keybit, which was mitigated by limiting the number of keys enable.

Reply:
750 works well on my system: gnome wayland with both an external keyboard and a laptop keyboard.

But I am ok with 300. Will change that.